content/blog/eng/a-redux-pattern.md

```diff
@@ -0,0 +1,203 @@
+draft: true
+share: true
+categories: [Design patterns, Javascript]
+title: "A Redux pattern"
+languageCode: "en-EN"
+type: "post"
+toc: true
+twitterImage: /images/a-redux-pattern/elevator.jpg
+---
+
+# Introduction
+
+There is inconsistency around the Redux community about how to use actions. Redux Toolkit documentation suggests the following approach:
+
+> [...] we recommend trying to treat actions more as "describing events that occurred", rather than "setters".
+
+Why should we treat actions as events rather than setters? Dan Abramov, founder of Redux, said that Redux doesn't reinvent event sourcing. It's up to people how to use it. It's clear that there isn't a well accepted approach about how to use actions.
+
+In this article I'll talk about a design pattern for Redux. I'll show you how it can help resolve some known problems. In particular, I'll try to find a way to describe how to treat actions. This we'll lead us to deal better with the asynchronicity and the mutation problems. The main goal is to get a consistent state for our application. To achieve that, we'll look for a predictable way to mutate the state of the application.
+
+# Actions
+
+Starting from the basics, a Redux action has the following signature:
+
+```
+export interface Action<T = any> {
+ type: T
+}
+```
+
+An action is a plain object. It has a type, and it's `any` by default. This definition is general and not sufficient. We don't like to use `any` as a type. It would be better to have something more specific. Redux Toolkit (RTK) is a package intended to be the standard way to write Redux logic. RTK provides a set of tools which help write Redux logic. Among these tools, there are more specific types and that's why RTK is going to be our main reference for our journey. An action has the following signature:
+
+```
+export declare type PayloadAction<P = void, T extends string = string, M = never, E = never> = {
+ payload: P;
+ type: T;
+} & ([M] extends [never] ? {} : {
+ meta: M;
+}) & ([E] extends [never] ? {} : {
+ error: E;
+});
+```
+
+An action is an object. It can have some optional properties such as `payload`. The `type` property is `string` by default. Documentation suggests keeping the `type` property as a string because it's easier to serialize it.
+
+# Commands
+
+We know the signature of an action and we're ready to go deeper. Let's consider a real-life example.
+
+_You're getting out of a building and ready to take the elevator. You're in a hurry and start to hit the call button many times. The hope is to make the wait shorter. Hitting the button multiple times doesn't make the elevator either coming to you faster or coming at all. Indeed, it may be occupied or even out of service. At this point, there are two possible scenarios happening. Good scenario: the elevator arrives and you leave the building. Bad scenario: the elevator doesn't arrive. Then you decide to go down stairs on foot (hopefully you're not at the last floor of Burj Khalifa)._
```

Comment:
At this point, there are two possible outcomes: ....

.. you go downstairs on foot.